sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala

```diff
@@ -2063,4 +2063,178 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
       )
     }
   }
+
+  def testFile(fileName: String): String = {
+    Thread.currentThread().getContextClassLoader.getResource(fileName).toString
+  }
+
+  test("json in UTF-16 with BOM") {
+    val fileName = "json-tests/utf16WithBOM.json"
+    val schema = new StructType().add("firstName", StringType).add("lastName", StringType)
+    val jsonDF = spark.read.schema(schema)
+      // The mode filters null rows produced because new line delimiter
+      // for UTF-8 is used by default.
```

Comment:
This is funny that we have a test for this.

Reply:
The test came from customer's use case, when we broke backward compatibility with previous versions by forcibly set input stream as UTF-8:  https://github.com/apache/spark/pull/20302 . You can see the test case in the PR  when jackson-json parser is not able to detect charset correctly.

Reply:
This test case checks if a JSON file encoded with UTF-16 can be parsed line by line with UTF-8 newlines, and then each line is parsed by Jackson. Your customer's usecase accidentally works and I think we never documented this behaviour so far.

Reply:
Could you point me out the place in docs where we strictly restricted input charset to UTF-8 only.

Reply:
Could you point out where we support other encodings?

Reply:
We declare that we are able to read JSON. According to the rfc7159 (8.1 Character Encoding):

```
   JSON text SHALL be encoded in UTF-8, UTF-16, or UTF-32.  The default
   encoding is UTF-8, and JSON texts that are encoded in UTF-8 are
   interoperable in the sense that they will be read successfully by the
   maximum number of implementations; there are many implementations
   that cannot successfully read texts in other encodings (such as
   UTF-16 and UTF-32).
```

Users can think that Spark can read json in charset different from UTF-8 because it SHALL do that according to the rfc, and we DON'T directly declare that jsons such encodings cannot be read successfully.

Reply:
@MaxGekk, see what happens in the test code here now. Lines are separated by a newline with UTF-8 and then the records are parsed by a different encoding.

Reply:
Also, this is where we need a decision, right? It already does not work correctly. Another option for a min fix to follow rfc7159 is to describe that we don't support other encodings for now, to be clear.

I approved https://github.com/apache/spark/pull/20614 only respecting/assuming that it causes an actual issue to some sites and the release was close (which is true I guess now).